artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/RefsOperation.java

```diff
@@ -175,6 +179,8 @@ protected void rollbackRedelivery(Transaction tx, MessageReference ref, long tim
    @Override
    public void afterCommit(final Transaction tx) {
       for (MessageReference ref : refsToAck) {
+         clearLingerRef(ref);
```

Comment:
You dont need to do this inside the iterator. You have the operation queue and only need be done one

Reply:
Refs maybe from different queues, E.g. one session creates two consumers, one subscribing queue1, the other subscribing queue2.

Reply:
The ref operation holds the queue its for.

Reply:
I think it just hold the queue which the first acked message belongs to. Following acked messages in this transaction maybe from other queues since a session might create consumers subscribing to different queues.

Reply:
Then there would be seperate ref ops in the tx

Reply:
There is only one ref ops for each tx. I've added a test to validate it.

Reply:
I stand corrected :) nice test :)